src/meta/src/barrier/mod.rs

```diff
@@ -253,94 +301,110 @@ where
             }
             let new_epoch = Epoch::now();
             assert!(new_epoch > state.prev_epoch);
-            let command_ctx = CommandContext::new(
-                self.fragment_manager.clone(),
-                self.env.stream_clients_ref(),
-                &info,
-                &state.prev_epoch,
-                &new_epoch,
-                command.clone(),
-            );
-
-            let mut notifiers = notifiers;
-            notifiers.iter_mut().for_each(Notifier::notify_to_send);
-            match self.run_inner(&command_ctx).await {
-                Ok(responses) => {
-                    // Notify about collected first.
-                    notifiers.iter_mut().for_each(Notifier::notify_collected);
-
-                    // Then try to finish the barrier for Create MVs.
-                    let actors_to_finish = command_ctx.actors_to_finish();
-                    unfinished.add(new_epoch.0, actors_to_finish, notifiers);
-                    for finished in responses.into_iter().flat_map(|r| r.finished_create_mviews) {
-                        unfinished.finish_actors(finished.epoch, once(finished.actor_id));
-                    }
-
-                    state.prev_epoch = new_epoch;
-                }
-                Err(e) => {
-                    notifiers
-                        .into_iter()
-                        .for_each(|notifier| notifier.notify_collection_failed(e.clone()));
-                    if self.enable_recovery {
-                        // If failed, enter recovery mode.
-                        let (new_epoch, actors_to_finish, finished_create_mviews) =
-                            self.recovery(state.prev_epoch, Some(command)).await;
-                        unfinished = UnfinishedNotifiers::default();
-                        unfinished.add(new_epoch.0, actors_to_finish, vec![]);
-                        for finished in finished_create_mviews {
-                            unfinished.finish_actors(finished.epoch, once(finished.actor_id));
-                        }
 
-                        state.prev_epoch = new_epoch;
-                    } else {
-                        panic!("failed to execute barrier: {:?}", e);
+            let envconnet = self.env.clone();
+            let fragment_manager = self.fragment_manager.clone();
+            let metrics = self.metrics.clone();
+            let barrier_send_tx_clone = barrier_send_tx.clone();
+            let heap = self.epoch_heap.clone();
+            let hummock_manager = self.hummock_manager.clone();
+
+            tokio::spawn(async move {
```

Comment:
How about using multiple coroutines instead of multiple threads here? 🤔

We can collect the futures and await them without thread-overprovisioning, where the latency of RPC is also overlapped.
```rust
futures.push(async move{ ... })
...
let res = select_all(futures).await;
```

If coroutine works, there would be no data race on `env` and `epoch_heap`, so `RwLock` could be mitigated.
```rust
    env: Arc<MetaSrvEnv<S>>,

    epoch_heap: Arc<BinaryHeap<u64>>,
```

Reply:
okoko , I will do it after I complete the basic functions.